### ISSUE TITLE
Getting an error when unpacking file. 

### ISSUE THREAD

**Issue by Red1860** [baraklevy20/SCP-Unpacker#2]
Hi, I am getting an error when unpacking scp files. 

```
Error: Unsupported flag: CHACHA20
    at processFlags (/Users/admin/SCP-Unpacker/src/scp.js:8:11)
    at unpack (/Users/admin/SCP-Unpacker/src/scp.js:52:27)
    at /Users/admin/SCP-Unpacker/index.js:30:19
    at Array.forEach (<anonymous>)
    at main (/Users/admin/SCP-Unpacker/index.js:27:18)
    at Object.<anonymous> (/Users/admin/SCP-Unpacker/index.js:43:1)
    at Module._compile (node:internal/modules/cjs/loader:1101:14)
    at Object.Module._extensions..js (node:internal/modules/cjs/loader:1153:10)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.Module._load (node:internal/modules/cjs/loader:822:12)
Unpack time - 8ms

**Comment by baraklevy20** [comment]
Oh, that's great I've been looking for files that use CHACHA20. Can you upload that file?

**Comment by Red1860** [comment]
Sure - 

[c86ed78b-5ad6-4fe1-9974-f9c0d1c77fab.sc.zip](https://github.com/baraklevy20/SCP-Unpacker/files/7654092/c86ed78b-5ad6-4fe1-9974-f9c0d1c77fab.sc.zip)
[3c7f6738-1e34-49d0-9454-9f3cfb7e7ace_skin_20211129173737_ios.scp.zip](https://github.com/baraklevy20/SCP-Unpacker/files/7654093/3c7f6738-1e34-49d0-9454-9f3cfb7e7ace_skin_20211129173737_ios.scp.zip)

**Comment by baraklevy20** [comment]
Hi mate, I'm sorry but I can't seem to get the unpacker to decrypt the file. It seems to be using a custom encryption that I can't crack down. So for now I'm leaving this open and may come back to it later. I suggest you try downloading the APK (Android) version, they don't use encrypted SCP yet AFAIK.

**Comment by Red1860** [comment]
No problem, I will try the android version and when you update it let me know and I'll try it out.

**Comment by Red1860** [comment]
Hey, I got my hands on the android version and the files are the same and so is the error message. ¯\_(ツ)_/¯

**Comment by baraklevy20** [comment]
I've looked deeper into it and I'm pretty sure it's actually impossible to decrypt the file using only the SC file.

In order to decrypt the file, we need a key and a nonce. The SC file contains the **hash** of the key + nonce, but not the key and nonce themselves. The game stores the key and the nonce somewhere in `libg.so` (that's the main game library), and uses the hash in the SC file to access it (a lookup table).

In other words, in order to read the SC file, you need to read `libg.so`, find where it stores its encryption keys and nonces, and use the hash to access it. Then you can decrypt the file using the key and nonce.
The problem is that reading the `libg.so` to extract the keys is a much more difficult task, and way beyond my knowledge. I will not be supporting it. Hopefully, someone else will figure something out.

**Comment by baraklevy20** [comment]
@Red1860 Which game/version were you trying to unpack?

**Comment by Red1860** [comment]
@baraklevy20 Same, Clash of Clans latest version. You won't find these files in the apk. When you install the game it downloads additional data directly from supercell's server in a cache folder. These files are from that folder.

**Comment by baraklevy20** [comment]
@Red1860 I've managed to get into the cache folder, but I still can't find the files. Can we talk on discord? Barak#7575

**Comment by Red1860** [comment]
@baraklevy20 Sure I'll send you a friend request.

**Comment by Red1860** [comment]
> I've looked deeper into it and I'm pretty sure it's actually impossible to decrypt the file using only the SC file.
> 
> In order to decrypt the file, we need a key and a nonce. The SC file contains the **hash** of the key + nonce, but not the key and nonce themselves. The game stores the key and the nonce somewhere in `libg.so` (that's the main game library), and uses the hash in the SC file to access it (a lookup table).
> 
> In other words, in order to read the SC file, you need to read `libg.so`, find where it stores its encryption keys and nonces, and use the hash to access it. Then you can decrypt the file using the key and nonce. The problem is that reading the `libg.so` to extract the keys is a much more difficult task, and way beyond my knowledge. I will not be supporting it. Hopefully, someone else will figure something out.

Hi, is is possible to unzip the files once we get the nonce and the key?

**Comment by baraklevy20** [comment]
> > I've looked deeper into it and I'm pretty sure it's actually impossible to decrypt the file using only the SC file.
> > In order to decrypt the file, we need a key and a nonce. The SC file contains the **hash** of the key + nonce, but not the key and nonce themselves. The game stores the key and the nonce somewhere in `libg.so` (that's the main game library), and uses the hash in the SC file to access it (a lookup table).
> > In other words, in order to read the SC file, you need to read `libg.so`, find where it stores its encryption keys and nonces, and use the hash to access it. Then you can decrypt the file using the key and nonce. The problem is that reading the `libg.so` to extract the keys is a much more difficult task, and way beyond my knowledge. I will not be supporting it. Hopefully, someone else will figure something out.
> 
> Hi, is is possible to unzip the files once we get the nonce and the key?

I believe so, yeah.

**Comment by Red1860** [comment]
I've sent you a file on Discord

**Comment by baraklevy20** [comment]
Xsalsa20 is now supported so I'm closing this issue.
